### PR TITLE
Royalty Rate Schedule & Royalty Supplemental Payments [v3.11.25]

### PR DESCRIPTION
**Pull request by softwareengineerprogrammer**
# Description

1. **Implements https://github.com/NatLabRockies/GEOPHIRES-X/issues/466?title=Custom+royalty+schedules via new `Royalty Rate Schedule` parameter.**
    1. `Royalty Rate Schedule` tooltip text: "A schedule DSL string defining the royalty rate for each year of the project, starting at the first year of plant operations (Year 1, post-construction). Syntax: `<rate> * <years>, <rate> * <years>, ..., <terminal_rate>`. For example "0.0175 * 10, 0.035" means 1.75% for 10 years then 3.5% thereafter. Note: Providing both Royalty Rate and Royalty Rate Schedule is invalid and will result in an error."
    1. Example: https://beta.gtp.scientificwebservices.com/geophires/?geophires-example-id=example_SAM-single-owner-PPA-4b (also includes supplemental payments)
        1. `Royalty Rate Schedule, 0.025 * 5, 0.05 * 5, 0.075`
        1. image
    1. Cashflow screenshot: image
3. **Adds `Royalty Supplemental Payments` parameter to support lease/rental/etc. payments**
    1. `Royalty Supplemental Payments` tooltip text: "A schedule DSL string defining the royalty supplemental payments for each year of the project. Unlike the Royalty Rate Schedule which begins at operations, this schedule begins at the first year of project construction. Syntax: `<amount> * <years>, <amount> * <years>, ..., <terminal_amount_per_year>`. For example "1 * 2, 0.25" means $1M for 2 years (e.g., during construction) then $250k/year thereafter." 
    1. Example: https://beta.gtp.scientificwebservices.com/geophires/?geophires-example-id=example_SAM-single-owner-PPA-5
        1. `Royalty Supplemental Payments, 1.5, 0.1, -- $1.5M first year option payment and $100k/yr land lease payment thereafter`
        2. image
    3. Cashflow CSV: https://github.com/softwareengineerprogrammer/GEOPHIRES/blob/royalty-schedules-3/tests/examples/example_SAM-single-owner-PPA-5_cash-flow.csv
    4. Cashflow screenshot: image

## Documentation Updates

1. **SAM Economic Models: parameter mappings & updated royalties section:
[sam-em-mcy-docs-updated-for-royalty-schedules-and-supplemental-payments_2026-03-09.pdf](https://github.com/user-attachments/files/25848427/sam-em-mcy-docs-updated-for-royalty-schedules-and-supplemental-payments_2026-03-09.pdf)**
    1. image
    3. image
1. **SAM-EM Multiple Construction Years: updated for royalty supplemental payments during construction years: [sam-em-docs-updated-royalties-sections_2026-03-09.pdf](https://github.com/user-attachments/files/25848390/sam-em-docs-updated-royalties-sections_2026-03-09.pdf)**
    1. image

# Testing & Verification

1. Unit tests added:
    1. [`test_expand_schedule`](https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/changes#diff-c4c05331de6e24956eaa84553b43d8eaabb27309fc6aaa9faf2b2934c0fbcc0fR9)
    1. [`test_royalty_supplemental_payments`](https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/changes#diff-894b10b737d88bd3c08bc3eaf69146781055586f575a582c09b5911feb891b38R1179)
    1. [`example_SAM-single-owner-PPA-4b`](https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/changes#diff-3856580731ef9dd8a6774d12f4b581733cfd0063dd626318406d0e17fb3459d9)
1. Unit tests updated:
    1. [`test_royalty_rate_schedule`](https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/changes#diff-894b10b737d88bd3c08bc3eaf69146781055586f575a582c09b5911feb891b38R1018)
    2. [`test_royalty_schedule_ignores_rate_modifier_params`](https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/changes#diff-894b10b737d88bd3c08bc3eaf69146781055586f575a582c09b5911feb891b38R1119)
    3. [`test_expand_schedule_dsl`](https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/changes#diff-894b10b737d88bd3c08bc3eaf69146781055586f575a582c09b5911feb891b38R1119)
    4. [`example_SAM-single-owner-PPA-5`](https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/changes#diff-8bf8035525dcb8cb1f516ab6174d25f4fb43e04364a23937d443e5e55e2afb4a)
1. Manual testing:
    1. Tested in consumer application
    6. Tested in web interface

---

## TODO
1. [X] SAM SME design review: discussed in GEOPHIRES sync
1. [ ] 🚧 Web interface support - deployed in beta, pending SME review before merge to main and deployment to prod
    1. https://beta.gtp.scientificwebservices.com/geophires/?geophires-example-id=example_SAM-single-owner-PPA-4b
    1. https://beta.gtp.scientificwebservices.com/geophires/?geophires-example-id=example_SAM-single-owner-PPA-5
1. [ ] 🚧 CHANGELOG entry (pending SME review and merge into fork main branch): See draft CHANGELOG entry in change checklist below
    1. [X] Note change in `Royalty Holder Average Annual Revenue` calculation - now lower because it accounts for construction years (and not just operational years)

### Done

1. [X] Support `Royalty Supplemental Payments`
    1. [X] During Construction in EconomicsSamPreRevenue.py: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/b7adc72bfb7ed4bc1fc424d195e0923a7743745c
        1. [X] Cash flow line item: `Royalty supplemental payments [construction] ($)`
        1. [X] Include construction years supplementary payments in capital costs results: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/b590d1f218cd7dbb6e7af123ee96c2160cb6dbd6 
    1. [X] During Operations: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/898f8df71a2a5ca653ac1e5a7e200d4bc6d1a99c, https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/898f8df71a2a5ca653ac1e5a7e200d4bc6d1a99c, https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/ec303b852a9df36c11bca4011410f51b87bb1008
        1. [X] Cash flow line item: Deferred to new tracking issue: https://github.com/NatLabRockies/GEOPHIRES-X/issues/469
        1. [X] Include in result OPEX: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/898f8df71a2a5ca653ac1e5a7e200d4bc6d1a99c
            1. [X] Average Annual Royalty Cost (~~excludes supplementary payments during construction years~~): https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/a258deaa1b7a2cbdb86d818a5ea70cfffada2d3a
    1. [X] Update Royalty Holder economics outputs
        1. [X] Include supplementary payments during construction years in royalty holder NPV, average revenue, and total revenue: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/a258deaa1b7a2cbdb86d818a5ea70cfffada2d3a
        1. [X] ~~Add output(s) for Royalty Supplemental Payments (total revenue, maybe others)~~ - deferred for now, could be added in future follow-up
    1. [X] Negative value - decide on support and add appropriate unit test: treat negative values as invalid and raise exception: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/864b9cb782a0331e406af4938d7e9f83dd5a95f2
1. [X] Validate mutually exclusive royalties params (schedule vs. rate-based): https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/0cc9a55fcd59a4a370bf817968cc9a2d4a33e3ab, https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/e05066c319ed52b54f4e89b705e7407fde2772b0
    1. Potential follow-up: maybe rate escalation could be applied to terminal schedule value in the future; left door open for this in https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/e05066c319ed52b54f4e89b705e7407fde2772b0
1. [X] Unit test that extracts rate-based royalties from cash flow, re-runs as schedule, and verifies results are identical: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/changes#diff-894b10b737d88bd3c08bc3eaf69146781055586f575a582c09b5911feb891b38R1078-R1116
1. [X] Add unit test for Royalty Supplemental Payments without production-based royalties: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/17f29bb80d2560158cd33ff4737e2fef7f386a68
1. [X] Documentation updates
    1. [X] tooltip texts - clarification of production-based vs. supplemental royalty payments (some marked TODO in diff)
        1. [X] https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139#discussion_r2866699931: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139#discussion_r2885566693
        1. [X] https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/e133b0bd0921b9acfedc16b5ac50bc747dae4f6b, https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/cf6e70ccf2131f11f09804ff912bc8a99c282748
        1. [X] \<other TODOs marked in diff\>: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/cf6e70ccf2131f11f09804ff912bc8a99c282748, https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/b5c287f89706b3e199f1f823d9c5626e588229d7
    1. [X] SAM Economic Models Royalties documentation (https://softwareengineerprogrammer.github.io/GEOPHIRES/SAM-Economic-Models.html#royalties): https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/a14a8ab00bd43cc7fe65e18f87a310d7e67a39f1, https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/607395cfbe4255399317405c892afc34532cc12d, https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/d9d40b9c76a95466ba82dda9e25501f75ec4308b
    1. [x] Multiple construction years cash flow (https://softwareengineerprogrammer.github.io/GEOPHIRES/SAM-EM_Multiple-Construction-Years.html)
        1. [X] text: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/766c1ca8fffabbed2a7040f1596ac9518b0b5c8b
        1. [X] screenshots: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/0e0391a64fb28dd98ed39463b16a8c4c3a335d05
        1. [X] update example 5 to include royalty supplemental payments: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/0e0391a64fb28dd98ed39463b16a8c4c3a335d05
        1. [X] diagram: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/0c1f0b1ddb18237797ec8f0b55a960f1267d2e96
            1. image
1. [X] Consolidate `'*'` param read matching logic: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/0dec59d0f0c3eed52f307a5952b42ad82fbf11f0
1. [X] capex results category must include supplemental payments total: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/029bc3a0f3a14fcd31b94660e705f98574b042c9, https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/e14c83a374f13c1534565494fc7a6568b8f0afb1
1. [X] `example_SAM-single-owner-PPA-4b` README list entry: https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/1591e74ae826b0d695af942a345a16e7e2623afb

---

**Change Checklist**:

Observe the following checklist items if your change includes a new version (see [versioning notes](https://github.com/NREL/GEOPHIRES-X/blob/main/CHANGELOG.rst#versioning-notes) and [version management docs](https://github.com/NREL/GEOPHIRES-X/blob/main/CONTRIBUTING.rst#version-management)).

Before merging:
1. [ ] Add a CHANGELOG entry ([example](https://github.com/NREL/GEOPHIRES-X/commit/15e10e5c3c08633a16645e41dfb7e8131805483f)). RST Template:
```restructuredtext
M.m.P: `Royalty Rate Schedule & Royalty Supplemental Payments <https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139>`__ | `release <https://github.com/NREL/GEOPHIRES-X/releases/tag/vM.m.P>`__ |  `documentation <https://softwareengineerprogrammer.github.io/GEOPHIRES/SAM-Economic-Models.html#royalties>`__ | **Changed:** The calculation for `Royalty Holder Average Annual Revenue` now averages over the entire project timeline (construction + operations) rather than just the operational phase. For projects with multi-year construction periods, this will result in a lower average value than in previous versions. See documentation for further details.
```

After merging:
1. [ ] Push tag(s) to GitHub: `git push origin refs/tags/vM.m.P`
1. [ ] Post a change announcement to the [GEOPHIRES Change Announcements Discussion](https://github.com/NREL/GEOPHIRES-X/discussions/428) ([example](https://github.com/NREL/GEOPHIRES-X/discussions/428#discussioncomment-14593405)). Markdown template:
```markdown
M.m.p: [Royalty Rate Schedule & Royalty Supplemental Payments](https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139) | [release](https://github.com/NREL/GEOPHIRES-X/releases/tag/vM.m.P)
```

---

*Self-reviewed in https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139*

### CODE REVIEW

File: tests/examples/example_SAM-single-owner-PPA-4.out [hunk 1]
```diff
@@ -439,5 +441,5 @@ Interest earned on reserves ($)                  	0            	0          	0
                                 ***EXTENDED ECONOMICS***
 
       Royalty Holder NPV:                                   50.59 MUSD
-      Royalty Holder Average Annual Revenue:                 4.40 MUSD/yr
+      Royalty Holder Average Annual Revenue:                 4.19 MUSD/yr
```
Comment (softwareengineerprogrammer) [inline tests/examples/example_SAM-single-owner-PPA-4.out]: TODO check that tooltip text reflects period this is calculated for (construction +Ops)

File: tests/examples/example_SAM-single-owner-PPA-4.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# Example: SAM Single Owner PPA Economic Model: 50 MWe with Royalties
+ # Example: SAM Single Owner PPA Economic Model: 50 MWe with Royalties
```
Comment (softwareengineerprogrammer) [inline tests/examples/example_SAM-single-owner-PPA-4.txt]: Remove extraneously added space

File: tests/examples/example_SAM-single-owner-PPA-4.out (hunk 1 above)
Comment (softwareengineerprogrammer) [reply tests/examples/example_SAM-single-owner-PPA-4.out]: Resolved by https://github.com/softwareengineerprogrammer/GEOPHIRES/pull/139/commits/6c5a5e808d251cecc4ad61e2072e7363d4f8c32d
